server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java

```diff
@@ -0,0 +1,179 @@
+            super(context);
+            this.fileInfo = fileInfo;
+            this.fileUploadListener = fileUploadListener;
+        }
+
+        @Override
+        public void run() {
+            ActionRunnable.run(fileUploadListener, () -> fileSnapshotter.accept(context, fileInfo)).run();
+        }
+
+        @Override
+        public short priority() {
+            return 2;
+        }
+    }
+
+    public ShardSnapshotWorkers(
+        final int maxWorkers,
+        final Executor executor,
+        final Consumer<SnapshotShardContext> shardSnapshotter,
+        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
+    ) {
+        assert maxWorkers > 0;
+
+        logger.info("starting shard snapshot worker pool of max size {}", maxWorkers);
+        this.maxWorkers = maxWorkers;
+        this.executor = executor;
+        this.shardSnapshotter = shardSnapshotter;
+        this.fileSnapshotter = fileSnapshotter;
+    }
+
+    public void enqueueShardSnapshot(final SnapshotShardContext context) {
+        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
+        synchronized (mutex) {
+            snapshotTasks.add(new ShardSnapshotTask(context));
+            ensureEnoughWorkers();
+        }
+    }
+
+    public void enqueueFileUpload(
+        final SnapshotShardContext context,
+        final FileInfo fileInfo,
+        final ActionListener<Void> fileUploadListener
+    ) {
+        logger.trace(
+            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
+            context.snapshotId(),
+            context.indexId(),
+            fileInfo.name()
+        );
+        synchronized (mutex) {
+            snapshotTasks.add(new FileSnapshotTask(context, fileInfo, fileUploadListener));
+            ensureEnoughWorkers();
+        }
+    }
+
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
+
+        int workersToCreate = Math.min(maxWorkers - workerCount, snapshotTasks.size());
+        if (workersToCreate > 0) {
+            logger.debug("starting {} shard snapshot workers", workersToCreate);
+        }
+        while (workersToCreate > 0) {
+            workerCount++;
+            startWorker(UUIDs.base64UUID());
+            workersToCreate--;
+        }
+        logger.debug("worker pool size is {}", workerCount);
+    }
+
+    private void workerDone() {
+        synchronized (mutex) {
+            assert workerCount > 0;
+
+            workerCount--;
+            ensureEnoughWorkers();
+        }
+    }
+
+    public int size() {
+        return workerCount;
+    }
+
+    private void startWorker(final String workerId) {
+        logger.debug("starting snapshot pool worker {}", workerId);
+        executor.execute(() -> {
+            try {
+                while (true) {
```

Comment:
I wonder if this should instead respawn to the back of the queue when there is more work, allowing other work to use the snapshot thread pool.

I suggest something like:

`startWorker` takes the `ShardSnapshotTask` as argument and executes it. It checks the queue afterwards and respawns (`executor.execute`) if there is more work to do - with the new task as argument.

We only queue excess tasks. I.e., if there is budget in `workerCount`, we always call `startWorker`, otherwise we queue it. We recheck `workerCount` after queueing to catch the edge case of the last worker exiting - and spawns a new task in that case (polling out the added task from the queue).

Reply:
@henningandersen Thanks for the clarification. TBH I think we misunderstood [your original suggestion](https://github.com/elastic/elasticsearch/pull/88209#issuecomment-1177696199)! What you are suggesting here makes way more sense that what I have in the PR! :-D  [I was wondering](https://github.com/elastic/elasticsearch/pull/88209#discussion_r925487903) whether we'd need to at least not use the same number of workers as the threadpool size since the workers would keep the threadpool busy as long as there are tasks in the queue!

Putting the lock-free part aside, I've tried out a draft of your suggestion, please see `ShardSnapshotWorkers.java` again. I've done some renaming as well, I guess the name worker doesn't have much meaning anymore. It's more about controlling max concurrent snapshot tasks and the spawning order. Is that correct?

Please let me know if this is in the right direction. I'll keep addressing the rest of the comments.

> We only queue excess tasks. I.e., if there is budget in workerCount, we always call startWorker, otherwise we queue it. We recheck workerCount after queueing to catch the edge case of the last worker exiting - and spawns a new task in that case (polling out the added task from the queue).

Not sure I completely understand this part. We'd have to always queue first, so that we run the tasks in the priority we want.

Reply:
> I think we misunderstood https://github.com/elastic/elasticsearch/pull/88209#issuecomment-1177696199

I think you got it just right, I think I educated my thinking here based on your work.

> Not sure I completely understand this part. We'd have to always queue first, so that we run the tasks in the priority we want.

I do not think this is important. It will be right most of the time if not all the time (depending on details in the code). Let me take a look at the change you did.

Reply:
I think the version you have could work out.

Anyway, I wanted to illustrate the lock-free approach. I think this preserves order just as well as the mutex one. Notice that if there is free capacity we always execute. We only queue if there is not. Meaning the queue is empty if there is free capacity. This is not entirely true in the mutex variant (but could be modified there too).

I think I prefer the lock-free version, but I'd also be ok with the mutex version (have comments for it though).

```
public class ShardSnapshotWorkers {
    private static final Logger logger = LogManager.getLogger(ShardSnapshotWorkers.class);

    private final int maxRunningTasks;
    private final Object mutex = new Object();
    private AtomicInteger runningTasksCount = new AtomicInteger();
    private final BlockingQueue<SnapshotTask> snapshotTasks = new PriorityBlockingQueue<>();
    private final Executor executor;
    private final Consumer<SnapshotShardContext> shardSnapshotter;
    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;

    abstract class SnapshotTask implements Comparable<SnapshotTask>, Runnable {
        protected final SnapshotShardContext context;

        SnapshotTask(SnapshotShardContext context) {
            this.context = context;
        }

        public abstract short priority();

        public SnapshotShardContext context() {
            return context;
        }

        @Override
        public final int compareTo(SnapshotTask other) {
            int res = context.snapshotId().compareTo(other.context.snapshotId());
            if (res != 0) {
                return res;
            }
            return Integer.compare(priority(), other.priority());
        }
    }

    class ShardSnapshotTask extends SnapshotTask {
        ShardSnapshotTask(SnapshotShardContext context) {
            super(context);
        }

        @Override
        public void run() {
            shardSnapshotter.accept(context);
        }

        @Override
        public short priority() {
            return 1;
        }

    }

    class FileSnapshotTask extends SnapshotTask {
        private final FileInfo fileInfo;
        private final ActionListener<Void> fileUploadListener;

        FileSnapshotTask(SnapshotShardContext context, FileInfo fileInfo, ActionListener<Void> fileUploadListener) {
            super(context);
            this.fileInfo = fileInfo;
            this.fileUploadListener = fileUploadListener;
        }

        @Override
        public void run() {
            ActionRunnable.run(fileUploadListener, () -> fileSnapshotter.accept(context, fileInfo)).run();
        }

        @Override
        public short priority() {
            return 2;
        }
    }

    public ShardSnapshotWorkers(
        final int maxRunningTasks,
        final Executor executor,
        final Consumer<SnapshotShardContext> shardSnapshotter,
        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
    ) {
        assert maxRunningTasks > 0;

        logger.info("starting shard snapshot worker pool of max size {}", maxRunningTasks);
        this.maxRunningTasks = maxRunningTasks;
        this.executor = executor;
        this.shardSnapshotter = shardSnapshotter;
        this.fileSnapshotter = fileSnapshotter;
    }

    public void enqueueShardSnapshot(final SnapshotShardContext context) {
        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
        runOrQueueTask(new ShardSnapshotTask(context));
    }

    public void enqueueFileSnapshot(
        final SnapshotShardContext context,
        final FileInfo fileInfo,
        final ActionListener<Void> fileUploadListener
    ) {
        logger.trace(
            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
            context.snapshotId(),
            context.indexId(),
            fileInfo.name()
        );
        FileSnapshotTask task = new FileSnapshotTask(context, fileInfo, fileUploadListener);
        runOrQueueTask(task);
    }

    private void runOrQueueTask(SnapshotTask task) {
        if (incrementRunningTasks()) {
            spawnTask(task);
        } else {
            snapshotTasks.add(task);
            if (incrementRunningTasks()) {
                pollAndSpawn();
            }
        }
    }

    private void pollAndSpawn() {
        SnapshotTask polledTask = snapshotTasks.poll();
        if (polledTask != null) {
            spawnTask(polledTask);
        } else {
            int decremented = runningTasksCount.decrementAndGet();
            assert decremented >= 0;
        }
    }

    private void spawnTask(SnapshotTask task) {
        executor.execute(() -> runOneTask(task));
    }

    private boolean incrementRunningTasks() {
        return runningTasksCount.getAndUpdate(v -> v < maxRunningTasks ? v + 1 : v) < maxRunningTasks;
    }

    // for testing
    int size() {
        synchronized (mutex) {
            return runningTasksCount.get();
        }
    }

    private void runOneTask(final SnapshotTask task) {
        try {
            logger.trace("running snapshot task {}", task); // TODO: toString()?
            task.run();
        } finally {
            pollAndSpawn();
        }
    }

}
```